src/uuid.cr

```diff
@@ -0,0 +1,208 @@
+      end
+    when 32 # Hexstring
+      16.times do |i|
+        string_has_hex_pair_at! value, i * 2
+        bytes[i] = value[i * 2, 2].to_u8(16)
+      end
+    when 45 # URN
+      raise ArgumentError.new "Invalid URN UUID format, expected string starting with \":urn:uuid:\"" unless value.starts_with? "urn:uuid:"
+      [9, 11, 13, 15, 18, 20, 23, 25, 28, 30, 33, 35, 37, 39, 41, 43].each_with_index do |offset, i|
+        string_has_hex_pair_at! value, offset
+        bytes[i] = value[offset, 2].to_u8(16)
+      end
+    else
+      raise ArgumentError.new "Invalid string length #{value.size} for UUID, expected 32 (hexstring), 36 (hyphenated) or 46 (urn)"
+    end
+
+    new(bytes, variant, version)
+  end
+
+  # Raises `ArgumentError` if string `value` at index `i` doesn't contain hex
+  # digit followed by another hex digit.
+  private def self.string_has_hex_pair_at!(value : String, i)
+    unless value[i, 2].to_u8(16, whitespace: false, underscore: false, prefix: false)
+      raise ArgumentError.new [
+        "Invalid hex character at position #{i * 2} or #{i * 2 + 1}",
+        "expected '0' to '9', 'a' to 'f' or 'A' to 'F'",
+      ].join(", ")
+    end
+  end
+
+  # Generates RFC 4122 v4 UUID.
+  #
+  # It is strongly recommended to use a cryptographically random source for
+  # *random*, such as `Random::Secure`.
+  def self.random(random = Random::Secure, variant = Variant::RFC4122, version = Version::V4)
+    new_bytes = uninitialized UInt8[16]
+    random.random_bytes(new_bytes.to_slice)
+
+    new(new_bytes, variant, version)
+  end
+
+  def self.empty
+    new(StaticArray(UInt8, 16).new(0_u8), UUID::Variant::NCS, UUID::Version::V4)
+  end
+
+  # Returns UUID variant.
+  def variant
+    case
+    when @bytes[8] & 0x80 == 0x00
+      Variant::NCS
+    when @bytes[8] & 0xc0 == 0x80
+      Variant::RFC4122
+    when @bytes[8] & 0xe0 == 0xc0
+      Variant::Microsoft
+    when @bytes[8] & 0xe0 == 0xe0
+      Variant::Future
+    else
+      Variant::Unknown
+    end
+  end
+
+  # Returns version based on RFC4122 format. See also `#variant`.
+  def version
+    case @bytes[6] >> 4
+    when 1 then Version::V1
+    when 2 then Version::V2
+    when 3 then Version::V3
+    when 4 then Version::V4
+    when 5 then Version::V5
+    else        Version::Unknown
+    end
+  end
+
+  # Returns 16-byte slice.
+  def to_slice
+    @bytes.to_slice
+  end
+
+  # Returns unsafe pointer to 16-bytes.
+  def to_unsafe
+    @bytes.to_unsafe
+  end
+
+  # Returns `true` if `other` string represents the same UUID, `false` otherwise.
+  def ==(other : String)
+    self == UUID.new other
+  end
+
+  # Returns `true` if `other` 16-byte slice represents the same UUID, `false` otherwise.
+  def ==(other : UUID)
+    to_slice == other.to_slice
+  end
+
+  def to_s(io : IO)
+    slice = to_slice
+
+    buffer = uninitialized UInt8[36]
+    buffer[8] = buffer[13] = buffer[18] = buffer[23] = '-'.ord.to_u8
+
+    buffer_ptr = buffer.to_unsafe
```

Comment:
I have gotten requests to explicitly remove this before.
I think having pointer manipulation is fine but I have a feeling if anyone is still following this PR they might have comments.

Also is there any reason for this or are you just changing this?

Reply:
The `buffer` returned by `String.new(size, &block)` *is* a pointer. This is no less safe.